AVR: Add AVR support (tooling, core, and tests)

VoltagedDebunked opened the pull request:
**Detailed description**

Adds AVR architecture support.

* Implemented AVR disassembler backend
* Integrated with `cstool`
* Added CMake and legacy Makefile support
* Code formatted with `clang-format`

**Test plan**

Installed and built this version.
Compiled example AVR programs using my own code.
Validated disassembly output with `cstool`.

VoltagedDebunked committed:
AVR: Add AVR support (tooling, core, and tests)

VoltagedDebunked committed:
Fix formatting in cstool

VoltagedDebunked committed:
Add avr to config.mk

VoltagedDebunked committed:
Add avr to Makefile (for legacy support)

VoltagedDebunked committed:
Add avr to cmake.sh

Rot127 reviewed (changes requested):
Please revert the formatting to the non-related files.

Also, please extend `cstest` and generate the python binding constants so the tests can run.

Rot127 commented:
Thanks for the new module!

I am a little bit confused though how you built it. It does look like some code is from LLVM. But it is not generated by Auto-Sync. Have you worked on it by hand?

I would highly recommend to use Auto-Sync to port the module. It gives you more consistent results and can be updated more easily. Also you would get the 32bit instructions. Currently it looks like only the 16bit instructions are supported.

VoltagedDebunked commented:
Alright, ill work on it. thanks for the feedback!

Yes, i have worked on it by-hand, i never really worked that much with llvm, i did use the provided modules, and autosync somewhy just didnt work, so i made it manually.

ill try doing autosync though. maybe fix it, somehow.

Rot127 commented:
> somewhy just didnt work, so i made it manually.

Can you please describe what the error was you ran into? If you use Python 3.13 it will be fixed with https://github.com/capstone-engine/capstone/pull/2705. Until then it is necessary to use Python 3.12 unfortunately. You can do this with an virtual environment with [uv](https://docs.astral.sh/uv/).

```
uv venv --python 3.12
uv pip install suite/auto-sync/
```

VoltagedDebunked committed:
Add tests, bindings, and reverse formatting for irrelevant files.

VoltagedDebunked commented:
alright, ill try that.

thanks  a lot - that definetly helped.

Rot127 commented:
Nice! Let me know if I can be of any help.

I highly advice you to read the refactor guide as well: https://github.com/capstone-engine/capstone/blob/next/suite/auto-sync/RefactorGuide.md

It should give you a good intro about everything. And if you think something is not clear or missing, please let me know. So it can be improved.

You can also join our [Telegram community channel](https://t.me/CapstoneEngine) and ask questions there. Some more people with refactor experience are in there as well and might help quicker if I am unavailable. 

> use Python 3.13 it will be fixed with https://github.com/capstone-engine/capstone/pull/2705.

The PR is stuck on a bug in tree-sitter. I pinged the maintainers about it and hope they resolve it quickly. Otherwise will add a work-around. Sorry, super annoying that it doesn't work with Python 3.13 out of the box.

VoltagedDebunked commented:
> Nice! Let me know if I can be of any help.
> 
> I highly advice you to read the refactor guide as well: https://github.com/capstone-engine/capstone/blob/next/suite/auto-sync/RefactorGuide.md...

alr, i will definetly do it, thanks a lot for the help - i will make sure to read it.

VoltagedDebunked commented:
Yeah i think im gonna stick to manual, llvm tblgen wont work for me. The python stuff doesnt do either, i tried uv, maybe ill try again later.

Rot127 commented:
Can you please open an issue what is not working so we can fix it.

VoltagedDebunked commented:
> Can you please open an issue what is not working so we can fix it.

of course! ill gather some quick info, and then i will post it on issues.

VoltagedDebunked commented:
yeah i dont think its an actual issue i think its just me being autistic :sob:

Rot127 commented:
> yeah i dont think its an actual issue i think its just me being autistic 😭

If I can help somehow let me know.

We haven't decided internally yet if we can accept a manual PR.
Having AVR would be super awesome for sure! But the problem with mostly hand-written code is, that it is really hard to maintain. Especially since there is the option of using Auto-Sync.

Without it we can't cross check against LLVM, because the code base is not the same.
Also AVR might get extensions in the future. And having a semi-automated update mechanism for it just speeds things up a lot.

VoltagedDebunked commented:
> > yeah i dont think its an actual issue i think its just me being autistic 😭
> 
> If I can help somehow let me know.
> 
> We haven't decided internall...

alright, ill try to tinker with it, and then try actual rewrite with the standard approach.

VoltagedDebunked commented:
nevermind, i dont think it works for me. it breaks my system's llvm since im on arch linux, auto-sync had to be rewritten by me just for it to not clean immediately after generating the c++ stuff, had to reclone the same repo numerous times, and really - i dont mind maintaining avr, i mean - hppa is manually implemented, so why not this one? its a simple 8-bit microcontroller, i think using a whole toolchain with complex syncs will be overkill. if you can, ill polish anything else, but using llvm, i dont really think of. and havent really used llvm tooling (except for clang and its tools) before. thanks for understanding.

VoltagedDebunked committed:
Add CSTest cases for AVR

Rot127 committed:
Add MC tests from LLVM.

notxvilka commented:
> nevermind, i dont think it works for me. it breaks my system's llvm since im on arch linux, auto-sync had to be rewritten by me just for it to not clean immediately after generating the c++ stuff, had to reclone the same repo numerous times, and really - i dont mind maintaining avr, i mean - hppa is manually implemented, so why not this one? its a simple 8-bit microcontroller, i think using a whole toolchain with complex syncs will be overkill. if you can, ill polish anything else, but using llvm, i dont really think of. and havent really used llvm tooling (except for clang and its tools) before. thanks for understanding.

HPPA is not in the mainstream LLVM. And for not breaking your system LLVM you can just use a docker container (or alternatives). Just my 2c.

VoltagedDebunked commented:
> > nevermind, i dont think it works for me. it breaks my system's llvm since im on arch linux, auto-sync had to be rewritten by me just for it to not clean immdiately after generating the c++ stuff, had to reclone the same repo numerous times, and really - i dont mind maintaining avr, i mean - hppa is manually implemented, so why not this one? its a simple 8-bit microcontroller, i think using a whole toolchain with complex syncs will be overkill. if you can, ill polish anything else, but using llvm, i dont really think of. and havent really used llvm tooling (except for clang and its tools) before. thanks for understanding.
> 
> HPPA is not in the mainstream LLVM. And for not breaking your system LLVM you can just use a docker container (or alternatives). Just my 2c.

still, id honestly just try and make it manual, i dont mind maintaining it, i dont have anything better to do honestly

Rot127 commented:
> still, id honestly just try and make it manual, i dont mind maintaining it, i dont have anything better to do honestly

Feel free to go ahead. Having AVR is nice one way or the other. Please merge https://github.com/VoltagedDebunked/capstone/pull/1 so you have a good test coverage to work with.

VoltagedDebunked committed:
Merge pull request #1 from Rot127/avr-mc-tests

Add MC tests from LLVM.

VoltagedDebunked commented:
Merged! thanks a lot for the help, ill test it now.

VoltagedDebunked commented:
alright i tested it, looks good, pretty much, it works just fine and i disassembled a bunch of avr binaries, using my own program, i installed the library locally, and then i built a test program, and it integrates quite solidly, no segfaults, and pretty accurate disassembly. i think its really close to a merge.

VoltagedDebunked committed:
Fix decoding parts (some were broken, like brne turned into brbc)

VoltagedDebunked commented:
alright, i got one major issue fixed:

Some encodings were incorrect. i fixed it by changing the register extraction.

also added support for X/Y/Z + instructions.

Rot127 committed:
Add fuzzing tests.

Rot127 commented on arch/AVR/AVRDecoder.c:
```diff
@@ -0,0 +1,575 @@
+
+static DecodeStatus DecodeBranch(MCInst *MI, uint16_t insn, uint64_t addr, const void *decoder)
+{
+    int16_t offset = (int16_t)(insn << 4) >> 4; // Sign extend 12-bit offset
```
Please implement a new helper function for sign extension in `MathExtras.h` like there already is for `SignExtend32`.

Rot127 commented on arch/AVR/AVRDecoder.c:
```diff
@@ -0,0 +1,575 @@
+
+static DecodeStatus DecodeBranchSR(MCInst *MI, uint16_t insn, uint64_t addr, const void *decoder)
+{
+    int8_t offset = (int8_t)(insn << 1) >> 1;  // Sign extend 7-bit offset
```
Also please add a SignExtend function.

Rot127 commented on arch/AVR/AVRDecoder.c:
```diff
@@ -0,0 +1,575 @@
+        mode = 1;
+    } else {
+        return MCDisassembler_Fail;
+    }
```
Replace with a switch

Rot127 commented on arch/AVR/AVRDecoder.c:
```diff
@@ -0,0 +1,575 @@
+{
+    unsigned Rd = (insn >> 4) & 0x1F;
+    unsigned ptr = 0;
+    unsigned mode = 0; // 0=normal, 1=post-inc, 2=pre-dec
```
No magic numbers here please. Add an enum and use it for the mode.
Also the addressing mode should also be in the details, right?

Rot127 commented on arch/AVR/AVRDisassembler.c:
```diff
@@ -0,0 +1,158 @@
+#ifdef CAPSTONE_HAS_AVR
+
+// Register decoder helpers
+static unsigned __attribute__((unused)) DecodeGPR8RegisterClass(MCInst *Inst, unsigned RegNo,
```
```suggestion
static unsigned DecodeGPR8RegisterClass(MCInst *Inst, unsigned RegNo,
```

Also below. At lease before we merge it.

Rot127 commented on arch/AVR/AVRDisassembler.c:
```diff
@@ -0,0 +1,158 @@
+{
+    // Sign extend 7-bit immediate
+    if (Imm & 0x40)
+        Imm |= 0xFFFFFF80;
```
Use function as above.

Rot127 commented on arch/AVR/AVRDisassembler.c:
```diff
@@ -0,0 +1,158 @@
+{
+    // Sign extend 12-bit immediate
+    if (Imm & 0x800)
+        Imm |= 0xFFFFF000;
```
Into function please.

Rot127 commented on arch/AVR/AVRDisassembler.c:
```diff
@@ -0,0 +1,158 @@
+static unsigned __attribute__((unused)) DecodeIORegister(MCInst *Inst, unsigned RegNo,
+                                uint64_t Address, const void *Decoder)
+{
+    if (RegNo > 63)
```
```suggestion
    if (RegNo >= AVR_REG_ENDING)
```

Also at the other places.

Rot127 commented on arch/AVR/AVRDisassembler.c:
```diff
@@ -0,0 +1,158 @@
+    uint16_t Insn = (Bytes[1] << 8) | Bytes[0];
+    *Size = 2;
+    
+    // Check for 32-bit instructions (JMP, CALL, LDS, STS)
+    if ((Insn & 0xFE0E) == 0x940C || // JMP
+        (Insn & 0xFE0E) == 0x940E || // CALL
+        (Insn & 0xFE0F) == 0x9000 || // LDS
+        (Insn & 0xFE0F) == 0x9200) { // STS
```
Use the function you defined above.

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+        if (RegNo == AVR_REG_X) return "X";
+        if (RegNo == AVR_REG_Y) return "Y";
+        if (RegNo == AVR_REG_Z) return "Z";
+        if (RegNo == AVR_REG_SP) return "SP";
```
Make them case labels please

Rot127 replied:
Actually, please us the lookup function from AVRMapping.c
Otherwise we have duplicate code.

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+    if (name) {
+        SStream_concat0(O, name);
+    } else {
+        SStream_concat(O, "UNKNOWN_REG_%u", RegNo);
```
```suggestion
        SStream_concat(O, "UNKNOWN_REG_%" PRIu32, RegNo);
```

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+            SStream_concat(O, "%"PRId64, imm);
+        else
+            SStream_concat(O, "$0x%"PRIx64, (uint64_t)imm);
+    }
```
Use `PrintUInt` functions

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+    
+    if (MCOperand_isImm(Op)) {
+        int64_t imm = MCOperand_getImm(Op);
+        SStream_concat(O, "0x%"PRIx64, (uint64_t)imm);
```
`printUInt`

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+            if (disp != 0) {
+                SStream_concat(O, "+%"PRId64, disp);
+            }
+        }
```
This code will never be reached. Because `MCOperand_isReg` was true.

Also please use `printUInt` functions

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+    
+    if (MCOperand_isImm(Op)) {
+        int64_t addr = MCOperand_getImm(Op);
+        SStream_concat(O, "0x%02"PRIx64, (uint64_t)addr);
```
printuint

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+        if (offset >= 0)
+            SStream_concat(O, ".+%"PRId64, offset * 2 + 2);
+        else
+            SStream_concat(O, ".%"PRId64, offset * 2 + 2);
```
printuint

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+    }
+    
+    // Set up instruction details for Capstone - with safety checks
+    if (MI->flat_insn && MI->flat_insn->detail) {
```
The `MI->flat_insn` check should be at the beginning of the function. But you can assume it is.
For the detail NULL check use `detail_is_set(MI)`. Also invert the condition and return early if detail is unset.

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+    
+    // Set up instruction details for Capstone - with safety checks
+    if (MI->flat_insn && MI->flat_insn->detail) {
+        cs_avr *avr = &(MI->flat_insn->detail->avr);
```
Add a helper in `Mapping.h` to get the AVR details. See `DEFINE_get_arch_detail`

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+    // Set up instruction details for Capstone - with safety checks
+    if (MI->flat_insn && MI->flat_insn->detail) {
+        cs_avr *avr = &(MI->flat_insn->detail->avr);
+        if (avr) {
```
This can be assumed.

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+            avr->op_count = 0;
+            
+            // Fill in operand details
+            for (unsigned i = 0; i < numOps && i < 3 && avr->op_count < 3; i++) {
```
Replace `3` with a macro.

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+            // Fill in operand details
+            for (unsigned i = 0; i < numOps && i < 3 && avr->op_count < 3; i++) {
+                MCOperand *Op = MCInst_getOperand(MI, i);
+                if (Op) {
```
Invert condition.

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+            for (unsigned i = 0; i < numOps && i < 3 && avr->op_count < 3; i++) {
+                MCOperand *Op = MCInst_getOperand(MI, i);
+                if (Op) {
+                    cs_avr_op *op = &avr->operands[avr->op_count];
```
Add helper with `DECL_get_detail_op` and use it here

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+                    } else if (MCOperand_isImm(Op)) {
+                        op->type = AVR_OP_IMM;
+                        op->imm = (int32_t)MCOperand_getImm(Op);
+                        avr->op_count++;
```
Add helper `DEFINE_inc_detail_op_count` and use it here.

Rot127 commented on arch/AVR/AVRInstPrinter.c:
```diff
@@ -0,0 +1,197 @@
+            
+            // Add instruction groups - with safety check
+            if (PrinterInfo) {
+                AVR_get_insn_id((cs_struct *)PrinterInfo, MI->flat_insn, OpCode);
```
`(cs_struct *)PrinterInfo` is no valid cast. `PrinterInfo` is of type `MCRegisterInfo *`

Rot127 commented on arch/AVR/AVRMapping.c:
```diff
@@ -0,0 +1,559 @@
+    }
+    
+    const char *name = id2name(reg_name_maps, sizeof(reg_name_maps)/sizeof(reg_name_maps[0]), reg);
+    if (!name) {
+        // Debug: print what register ID we couldn't find
+        // You can remove this later
+        static char unknown_reg[32];
+        snprintf(unknown_reg, sizeof(unknown_reg), "REG_%u", reg);
+        return unknown_reg;
+    }
```
It should just return NULL in that case. Because this leaks

Rot127 commented on arch/AVR/AVRMapping.c:
```diff
@@ -0,0 +1,559 @@
+        return NULL;  // Invalid register ID
+    }
+    
+    const char *name = id2name(reg_name_maps, sizeof(reg_name_maps)/sizeof(reg_name_maps[0]), reg);
```
```suggestion
    const char *name = id2name(reg_name_maps, ARR_SIZE(reg_name_maps), reg);
```

Rot127 commented on arch/AVR/AVRMapping.c:
```diff
@@ -0,0 +1,559 @@
+
+const char *AVR_insn_name(csh handle, unsigned int insn)
+{
+    return id2name(insn_name_maps, sizeof(insn_name_maps)/sizeof(insn_name_maps[0]), insn);
```
ARR_SIZE

Rot127 commented on arch/AVR/AVRMapping.c:
```diff
@@ -0,0 +1,559 @@
+
+const char *AVR_group_name(csh handle, unsigned int group)
+{
+    return id2name(group_name_maps, sizeof(group_name_maps)/sizeof(group_name_maps[0]), group);
```
ARR_SIZE

Rot127 commented on arch/AVR/AVRMapping.c:
```diff
@@ -0,0 +1,559 @@
+    case AVR_INS_IJMP:
+    case AVR_INS_EIJMP:
+    case AVR_INS_JMP:
+        if (insn->detail->groups_count < 8) {
```
No magic numbers please. Move the 8 into a macro def

Rot127 commented on arch/AVR/AVRMapping.c:
```diff
@@ -0,0 +1,559 @@
+    }
+    
+    // Set instruction groups based on the instruction type
+    switch (id) {
```
THe groups should be added earlier. Not in `AVR_get_insn_id`.

I would actually recomment to leave this function empty and just set everything in `AVR_printInstr`

Rot127 commented on arch/AVR/AVRModule.c:
```diff
@@ -0,0 +1,135 @@
+    MCRegisterInfo *mri;
+    mri = cs_mem_malloc(sizeof(*mri));
+    
+    MCRegisterInfo_InitMCRegisterInfo(mri, AVRRegDesc, sizeof(AVRRegDesc)/sizeof(AVRRegDesc[0]),
```
It seems like you don't use this profile. You can remove it, or am I missing something?

Rot127 commented on cstool/cstool.c:
Revert the formatting changes please.

Rot127 commented on cstool/cstool_aarch64.c:
revert

Rot127 replied:
Also for all the others.

Rot127 commented on cstool/cstool_avr.c:
```diff
@@ -0,0 +1,65 @@
+	if (avr->op_count)
+		printf("\top_count: %u\n", avr->op_count);
+
+	for (i = 0; i < avr->op_count; i++) {
```
Define `i` in the loop.

Rot127 commented on cstool/cstool_avr.c:
```diff
@@ -0,0 +1,65 @@
+			break;
+
+		case AVR_OP_IMM:
+			printf("\t\toperands[%d].type: IMM = 0x%x (%d)\n", i,
```
Use the `PRI` macros please

Rot127 commented on cstool/cstool_avr.c:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	// Print registers read/written by instruction
+	if (!cs_regs_access(handle, ins, regs_read, &regs_read_count,
```
Compare to `CS_ERR_OK` please

Rot127 commented on include/capstone/avr.h:
```diff
@@ -0,0 +1,231 @@
+
+#ifdef _MSC_VER
+#pragma warning(disable:4201)
+#pragma warning(disable:4214)
```
```suggestion
```

Rot127 commented on include/capstone/avr.h:
```diff
@@ -0,0 +1,231 @@
+    AVR_GRP_BRANCH_RELATIVE, // All relative branching instructions
+    
+    // Architecture-specific groups
+    AVR_GRP_ARITHMETIC,   // Arithmetic operations
```
```suggestion
    AVR_GRP_ARITHMETIC = 128,   // Arithmetic operations
```

Rot127 commented on include/capstone/avr.h:
```diff
@@ -0,0 +1,231 @@
+    AVR_OP_INVALID = 0, // Uninitialized/invalid operand
+    AVR_OP_REG,         // Register operand
+    AVR_OP_IMM,         // Immediate operand
+    AVR_OP_MEM,         // Memory operand
```
```suggestion
    AVR_OP_INVALID = CS_OP_INVALID, // Uninitialized/invalid operand
    AVR_OP_REG = CS_OP_REG,         // Register operand
    AVR_OP_IMM = CS_OP_IMM,         // Immediate operand
    AVR_OP_MEM = CS_OP_MEM,         // Memory operand
```

You need to include `cs_operand.h`

Rot127 commented on include/capstone/avr.h:
```diff
@@ -0,0 +1,231 @@
+// Instruction structure in AVR-specific format
+typedef struct cs_avr {
+    uint8_t op_count;   // Number of operands
+    cs_avr_op operands[3]; // Up to 3 operands per instruction
```
Move the `3` into a macro def

Rot127 commented on tests/MC/AVR/inst_adc_avr.txt.yaml:
```diff
@@ -0,0 +1,40 @@
+    input:
+      bytes: [ 0x0f, 0x1c ]
+      arch: "CS_ARCH_AVR"
+      options: [ "avr" ]
```
You can keep the options empty.

Rot127 commented on tests/MC/AVR/inst_adc_avr.txt.yaml:
```diff
@@ -0,0 +1,40 @@
+  -
+    input:
+      bytes: [ 0x0f, 0x1c ]
+      arch: "CS_ARCH_AVR"
```
`CS_ARCH_AVR` must be added to `test_mapping.h`. Don't forget to sort the table!

Rot127 reviewed (changes requested):
Thanks a lot for all the work!

Please make the tests run (see https://github.com/capstone-engine/capstone/pull/2726#discussion_r2133968195) and fix failing ones, and don't forget to add the fuzzing tests from https://github.com/VoltagedDebunked/capstone/pull/2 please

VoltagedDebunked commented:
alright i will

VoltagedDebunked committed:
Update include/capstone/avr.h

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

VoltagedDebunked committed:
Update include/capstone/avr.h

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

VoltagedDebunked committed:
Update arch/AVR/AVRDisassembler.c

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

VoltagedDebunked committed:
Update arch/AVR/AVRInstPrinter.c

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

VoltagedDebunked committed:
Update arch/AVR/AVRDisassembler.c

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

VoltagedDebunked committed:
Update arch/AVR/AVRMapping.c

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

VoltagedDebunked committed:
Update include/capstone/avr.h

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

VoltagedDebunked commented:
[ 75%] Building C object CMakeFiles/capstone.dir/arch/AVR/AVRInstPrinter.c.o
[ 76%] Building C object CMakeFiles/capstone.dir/arch/AVR/AVRDecoder.c.o
/home/voltaged/Projects/capstone-test/arch/AVR/AVRDisassembler.c:23:17: error: ‘DecodeGPR8RegisterClass’ defined but not used [-Werror=unused-function]
   23 | static unsigned DecodeGPR8RegisterClass(MCInst *Inst, unsigned RegNo,
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [CMakeFiles/capstone.dir/build.make:1549: CMakeFiles/capstone.dir/arch/AVR/AVRDisassembler.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:97: CMakeFiles/capstone.dir/all] Error 2
make: *** [Makefile:156: all] Error 2

i need to add __attribute__((unused)) to resolve the error.

VoltagedDebunked commented:
same for the other branch, io, and other functions.

VoltagedDebunked committed:
Apply requested fixes

Rot127 commented:
> i need to add attribute((unused)) to resolve the error.

Can't you just remove the function? Or comment it out until it is needed? In either case, for developing it is fine if they are still there. Just in the final version there shouldn't be anything unused.

VoltagedDebunked commented:
oh, sure! ill do that.

VoltagedDebunked committed:
Remove unused functions with __attribute__((unused))

VoltagedDebunked committed:
Merge pull request #2 from Rot127/fuzzing-tests

Add fuzzing tests.

VoltagedDebunked commented:
Done! Tested, too

Rot127 commented:
> Done! Tested, too

The CI is red though. Please fix it.

VoltagedDebunked commented:
alright

VoltagedDebunked committed:
Fix CI by removing unused function

VoltagedDebunked commented:
can you launch the CI please? so i can see if its fixed

VoltagedDebunked commented:
are u serious vro i cant even view it cuz it cuts off cuz my wifi so bad are you serious bruh

notxvilka commented:
> are u serious vro i cant even view it cuz it cuts off cuz my wifi so bad are you serious bruh

A useful hack - you can download log archives, just press the "gear" icon in the top right corner and choose "Download log archives", e.g. in your case it would be https://github.com/capstone-engine/capstone/suites/39790166017/logs?attempt=2
Then you can check them locally without any hassle.

VoltagedDebunked commented:
thanks!

VoltagedDebunked commented:
closed this pull request for now. im sorry, i just cant get it right. maybe ill come back later.

Rot127 commented:
Ok, feel free to continue it whenever you find time.
